src/tp_evaluation_engine.hpp

```diff
@@ -0,0 +1,434 @@
+  // length.
+  void SetTransitionAndDerivativeMatricesToHaveBranchLength(const double branch_length);
+  // Prep temporary transposed transition matrix variable to use given branch length.
+  void SetTransitionMatrixToHaveBranchLengthAndTranspose(const double branch_length);
+
+ protected:
+  // Partial Vector for storing Likelihood scores.
+  PLVEdgeHandler likelihood_pvs_;
+  // Tree likelihoods matrix across all sites.
+  EigenMatrixXd log_likelihoods_;
+  // Branch length parameters for DAG.
+  DAGBranchHandler dag_branch_handler_;
+
+  // ** Substitution Model
+  // When we change from JC69Model, check that we are actually doing transpose in
+  // leafward calculations.
+  JC69Model substitution_model_;
+  Eigen::Matrix4d eigenmatrix_ = substitution_model_.GetEigenvectors().reshaped(4, 4);
+  Eigen::Matrix4d inverse_eigenmatrix_ =
+      substitution_model_.GetInverseEigenvectors().reshaped(4, 4);
+  Eigen::Vector4d eigenvalues_ = substitution_model_.GetEigenvalues();
+  Eigen::Vector4d stationary_distribution_ = substitution_model_.GetFrequencies();
+
+  // ** Temporaries
+  // Stores intermediate computations useful for calculation.
+  EigenVectorXd per_pattern_log_likelihoods_;
+  EigenVectorXd per_pattern_likelihoods_;
+  EigenVectorXd per_pattern_likelihood_derivatives_;
+  EigenVectorXd per_pattern_likelihood_derivative_ratios_;
+  EigenVectorXd per_pattern_likelihood_second_derivatives_;
+  EigenVectorXd per_pattern_likelihood_second_derivative_ratios_;
+  Eigen::Vector4d diagonal_vector_;
+  Eigen::DiagonalMatrix<double, 4> diagonal_matrix_;
+  Eigen::Matrix4d transition_matrix_;
+  Eigen::Matrix4d derivative_matrix_;
+  Eigen::Matrix4d hessian_matrix_;
+};
+
+class TPEvaluationEngineViaParsimony : public TPEvaluationEngine {
+ public:
+  TPEvaluationEngineViaParsimony(TPEngine &tp_engine, const std::string &mmap_path);
+
+  // Initialize Computation Engine.
+  virtual void Initialize();
+  // Update the Computation Engine after adding Node Pairs to the DAG.
+  virtual void UpdateAfterDAGAddNodePair(const NNIOperation &post_nni,
+                                         const NNIOperation &pre_nni,
+                                         std::optional<size_t> new_tree_id);
+  // Computes scores. Call after Initialize or any Update steps, and before
+  // GetTopTreeScores.
+  virtual void ComputeScores();
+  // Get the Top Tree from the DAG with the given edge.
+  virtual double GetTopTreeScoreWithEdge(const EdgeId edge_id) const;
+  // Get the Top Tree from the DAG containing the proposed NNI.
+  virtual double GetTopTreeScoreWithProposedNNI(const NNIOperation &post_nni,
+                                                const NNIOperation &pre_nni,
+                                                const size_t spare_offset = 0);
+
+  // ** Resize
+
+  // Resize Engine for modified DAG.
+  virtual void GrowEngineForDAG(std::optional<Reindexer> node_reindexer,
```

Comment:
I'm confused why this is marked virtual. Shouldn't it be marked override instead?

Reply:
Marking with virtual is needed when the enclosing class will be used as a base class to inherit from. Marking with override is used in the derived class, and is used to ensure that both the base and derived's functions are virtual. TPEvaluationEngine has no base, so it doesn't make sense to have `override` functions. Classes that inherit from TPEvaluationEngine can use override, to ensure the virtual-ness is preserved trough future refactors of the base class.

Reply:
Sorry, I didn't notice it's in a derived class. This is a good place to use `override`, but it's not incorrect as it currently is. Being virtual allows for further inheritance if needed, and the value of using `override` and `final` is not too great to consider here. I imagine those specifiers can be helpful in codebases with excessive inheritance levels.

Reply:
Thanks for the input, I switched it to `override`.